app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -31,14 +25,41 @@ def sso_callback # rubocop:disable Metrics/AbcSize
       end
     end
 
+    def devise_sign_in
+      # The `sign_in_and_redirect` will only proceed with the login if the account has been confirmed
+      # either through trust by the IdP or with a confirmation mail. Until then, a flash message
+      # notifies users about the missing email confirmation
+      sign_in_and_redirect @user, event: :authentication
+      # We use the configured OmniAuth camilization to include the user-facing name of the provider in the flash message
+      set_flash_message(:notice, :success, kind: OmniAuth::Utils.camelize(provider)) if is_navigational_format?
+      session['omniauth_provider'] = provider
+    end
+
+    def nbp
```

Comment:
It should work now. I found it a bit tricky to implement this (event though it is not a lot of code). I felt that the User.from_omniauth method does a lot of things at the same time (like overwriting user attributes with new ones from the IdP, creating a new user if necessary, adding a new IdP if the user already exists, ...). Maybe we should split this up? This would probably be more verbose but I think it could improve readability.

Reply:
Sure, I am fine splitting the method further. Feel free to do so, if you like. Would that also allow you to reduce some of the "duplication" still present in the `omniauth_callbacks_controller.rb` further? I haven't executed it (yet), but just from reading the code, it reads like a regular sign in will perform two queries for the `user` (once in the `nbp` method for the `@user` method and a second time for the `User.from_omniauth` method).

I know, you are probably not a big fan of that, but can we swap the "responsibilities" here? My idea would be to call `sso_callback` for every provider and then just check which provider we are dealing with (either through the `provider` method or through something like `action_name` if nothing else works)?

Reply:
It is not exactly what you proposed but I refactored ```omniauth_callbacks_controller.rb``` quite a bit and I think it is much more readable now. What do you think?
Unfortunately I broke the ```deauthorize``` method but we can probably look at that on Wednesday.

Reply:
Looks pretty solid already, I like that. And the refactoring is nice, thanks. 

- I would still be in favour of making `nbp` an alias of `sso_callback` (and the methods currently only differ in a few lines anyway).
- Maybe we can find some more elegant solution to all those `set_flash_message` calls (extracting some static parts)? It uses a lot of space with many repetitions, currently...
- The `deauthorize` might be broken because of the `@current_user` vs `current_user` confusion? The `set_current_user` will only work as part of the SAML redirection workflow, not when accessing the page regularly. But, to be honest, I haven't executed the code (just guessing).
- Another non-guess as I was trying something else: `<class:OmniauthCallbacksController>': undefined method `sso_callback' for class `Users::OmniauthCallbacksController' (NameError)`. It seems like we should first define the method before defining an alias (happened when starting the Rails console).
- The call `@current_user.identities << @user_identity` in `add_idp_to_existing_user` is (probably) already persisted, so that the following `@current_user.save` would not make any persistent changes (it will still return `true` since the validation passes, but it is unexpected not to do the actual saving).

Reply:
Thanks, looks good! I am resolving this conversation, since all above points have been addressed.